mu4/notation/internal/notationsearchservice.cpp

```diff
@@ -0,0 +1,175 @@
+    case ElementType::MEASURE: {
+        //!NOTE: the measure numbering in the service starts from zero
+        int measureIndex = data.value.toInt() - 1;
+
+        Ms::Measure* measure = globalContext()->currentNotation()->elements()->measure(measureIndex);
+        if (!measure) {
+            return;
+        }
+
+        globalContext()->currentNotation()->interaction()->select(measure, SelectType::SINGLE);
+        break;
+    }
+    case ElementType::PAGE: {
+        //!NOTE: the page numbering in the service starts from zero
+        int pageIndex = data.value.toInt() - 1;
+
+        Ms::Page* page = globalContext()->currentNotation()->elements()->page(pageIndex);
+        if (!page) {
+            return;
+        }
+
+        globalContext()->currentNotation()->interaction()->select(page, SelectType::SINGLE);
+        break;
+    }
+    default:
+        globalContext()->currentNotation()->interaction()->clearSelection();
+        break;
+    }
+}
+
+NotationSearchService::SearchData NotationSearchService::parse(const std::string& text) const
+{
+    SearchData data;
+    if (text.empty()) {
+        return data;
+    }
+
+    data = parseMeasureCommand(text);
+    if (data.isValid()) {
+        return data;
+    }
+
+    data = parsePageCommand(text);
+    if (data.isValid()) {
+        return data;
+    }
+
+    data = parseRehearsalMarkCommand(text);
+    return data;
+}
+
+NotationSearchService::SearchData NotationSearchService::parseMeasureCommand(const std::string& text) const
+{
+    SearchData data;
+
+    QString qtext = QString::fromStdString(text).toLower();
+    bool ok = false;
+    int measureIndex = qtext.toInt(&ok);
+
+    if (ok) {
+        data.elementType = ElementType::MEASURE;
+        data.value = QVariant::fromValue(measureIndex);
+    }
+
+    return data;
+}
+
+NotationSearchService::SearchData NotationSearchService::parsePageCommand(const std::string& text) const
+{
+    SearchData data;
+
+    QString qtext = QString::fromStdString(text).toLower();
+
+    int pageCodeLength = static_cast<int>(PAGE_CODE.length());
+    if (!qtext.startsWith(PAGE_CODE.c_str()) || qtext.size() <= pageCodeLength) {
+        return data;
+    }
+
+    if (!qtext[pageCodeLength].isNumber()) {
+        return data;
+    }
+
+    bool ok = false;
+    int pageIndex = qtext.mid(PAGE_CODE.size()).toInt(&ok);
+
+    if (ok) {
+        data.elementType = ElementType::PAGE;
+        data.value = pageIndex;
+    }
+
+    return data;
+}
+
+NotationSearchService::SearchData NotationSearchService::parseRehearsalMarkCommand(const std::string& text) const
+{
+    SearchData data;
+
+    QString qtext = QString::fromStdString(text).toLower();
+
+    int rehearsalMarkCodeLength = static_cast<int>(REHEARSAL_MARK_CODE.length());
+    if (qtext.startsWith(REHEARSAL_MARK_CODE.c_str()) && qtext.size() > rehearsalMarkCodeLength) {
+        data.elementType = ElementType::REHEARSAL_MARK;
+        data.value = qtext.mid(rehearsalMarkCodeLength);
```

Comment:
Probably need some explanation here
Did you moved it from MU3?

Reply:
I didn't copy it(almost) :)
I am getting command and value from string here.
if the string starts with a specific command then the value will start from commandLenght to the end of the string.